openpype/hosts/zbrush/api/communication_server.py

```diff
@@ -0,0 +1,744 @@
+    encode_request, encode_error, decode_msg, JsonRpcMsgTyp
+)
+from aiohttp_json_rpc.exceptions import RpcError
+
+from openpype.lib import emit_event
+from openpype.hosts.zbrush import ZBRUSH_HOST_DIR
+
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+class CommunicationWrapper:
+    # TODO add logs and exceptions
+    communicator = None
+
+    log = logging.getLogger("CommunicationWrapper")
+
+    @classmethod
+    def create_qt_communicator(cls, *args, **kwargs):
+        """Create communicator for Artist usage."""
+        communicator = QtCommunicator(*args, **kwargs)
+        cls.set_communicator(communicator)
+        return communicator
+
+    @classmethod
+    def set_communicator(cls, communicator):
+        if not cls.communicator:
+            cls.communicator = communicator
+        else:
+            cls.log.warning("Communicator was set multiple times.")
+
+    @classmethod
+    def client(cls):
+        if not cls.communicator:
+            return None
+        return cls.communicator.client()
+
+    @classmethod
+    def execute_zscript(cls, zscript):
+        """Execute passed goerge script in Zbrush."""
+        if not cls.communicator:
+            return
+        return cls.communicator.execute_zscript(zscript)
+
+
+class WebSocketServer:
+    def __init__(self):
+        self.client = None
+
+        self.loop = asyncio.new_event_loop()
+        self.app = web.Application(loop=self.loop)
+        self.port = self.find_free_port()
+        self.websocket_thread = WebsocketServerThread(
+            self, self.port, loop=self.loop
+        )
+
+    @property
+    def server_is_running(self):
+        return self.websocket_thread.server_is_running
+
+    def add_route(self, *args, **kwargs):
+        self.app.router.add_route(*args, **kwargs)
+
+    @staticmethod
+    def find_free_port():
+        with closing(
+            socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        ) as sock:
+            sock.bind(("", 0))
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            port = sock.getsockname()[1]
+        return port
+
+    def get_port(self):
+        if self.port is not None:
+            return self.port
+
+    def start(self):
+        self.websocket_thread.start()
+
+    def stop(self):
+        try:
+            if self.websocket_thread.is_running:
+                log.debug("Stopping websocket server")
+                self.websocket_thread.is_running = False
+                self.websocket_thread.stop()
+        except Exception:
+            log.warning(
+                "Error has happened during Killing websocket server",
+                exc_info=True
+            )
+
+
+class WebsocketServerThread(threading.Thread):
+    """ Listener for websocket rpc requests.
+
+        It would be probably better to "attach" this to main thread (as for
+        example Harmony needs to run something on main thread), but currently
```

Comment:
Harmony -> Zbrush.

Also, this shows me there's quite some re-use of code across OpenPype. Should the "websocket" logic maybe be shared more across the different host implementations. Not saying we should do that now. But if there's one unified approach to integrations working with websockets etc. we can document it nicely easing the development on each of them.